File: 2626-count-the-number-of-good-subarrays/2626-count-the-number-of-good-subarrays.go
func countGood(nums []int, k int) int64 {
	n := len(nums)
	ans := int64(0)    
	freq := make(map[int]int)
	pairs := 0
	left := 0
	for right, num := range nums {
		pairs += freq[num]
		freq[num]++	

		for pairs >= k {
			ans += int64(n-right)
			freq[nums[left]]--
			pairs -= freq[nums[left]];
			left++
		}
	}

	return ans
}

File: 2626-count-the-number-of-good-subarrays/2626-count-the-number-of-good-subarrays.cpp
using ll = long long;
class Solution {
public:
  long long countGood(vector<int> &nums, int k) {
    int n = nums.size();
    unordered_map<int, int> freq;
    ll ans = 0;
    ll pairs = 0;
    int left = 0;

    for (int right = 0; right < n; ++right) {
      pairs += freq[nums[right]];
      freq[nums[right]]++;

      while (pairs >= k) {
        ans += (n - right);
        freq[nums[left]]--;
        pairs -= freq[nums[left]];
        left++;
      }
    }

    return ans;
  }
};
